public/app/plugins/datasource/loki/queryUtils.ts

```diff
@@ -295,3 +318,106 @@ export function getStreamSelectorsFromQuery(query: string): string[] {
 
   return labelMatchers;
 }
+
+const PARTITION_LIMIT = 60;
+export function partitionTimeRange(range: TimeRange, timeShift = 1, unit: DurationUnit = 'm'): TimeRange[] {
+  const partition: TimeRange[] = [];
+  for (let from = dateTime(range.from), to = dateTime(from).add(timeShift, unit); to <= range.to && from < range.to; ) {
+    partition.push({
+      from,
+      to,
+      raw: { from, to },
+    });
+
+    /**
+     * The user can request for any arbitrary time range. If we receive one from too
+     * long ago, this will cause an extremely large loop which could break the app.
+     * Additionally, data retention is limited, so ranges that go above PARTIION_LIMIT
+     * will be ignored.
+     */
+    if (partition.length > PARTITION_LIMIT) {
+      return [range];
+    }
+
+    from = dateTime(to);
+    to = dateTime(to).add(timeShift, unit);
+
+    if (to > range.to) {
+      to = range.to;
+    }
+  }
+
+  return partition;
+}
+
+export function requestSupportsPartitioning(queries: LokiQuery[]) {
+  /*
+   * For now, we would not split when more than 1 query is requested.
+   */
+  if (queries.length > 1) {
+    return false;
+  }
+
+  /**
+   * Disable logs volume queries.
+   */
+  if (queries[0].refId.includes('log-volume-')) {
+    return false;
+  }
+
+  if (isLogsQuery(queries[0].expr)) {
+    return false;
+  }
+
+  if (isRateQuery(queries[0].expr)) {
+    return false;
+  }
+
+  return true;
+}
+
+export function combineResponses(currentResult: DataQueryResponse | null, newResult: DataQueryResponse) {
+  if (!currentResult) {
+    return newResult;
+  }
+
+  newResult.data.forEach((newFrame) => {
+    const currentFrame = currentResult.data.find((frame) => frame.name === newFrame.name);
+    if (!currentFrame) {
+      currentResult.data.push(newFrame);
+      return;
+    }
+    combineFrames(currentFrame, newFrame);
+  });
+
+  return currentResult;
+}
+
+function combineFrames(dest: DataQueryResponseData, source: DataQueryResponseData) {
```

Comment:
```suggestion
function combineFrames(dest: DataFrame, source: DataFrame) {
```

I think we are expecting `DataFrame`s here, right?

Reply:
Not really sure. I've been using this type inference:

![Uploading imagen.png…]()

Reply:
I think this is only working because of this:
https://github.com/grafana/grafana/blob/matyax/range-splitting/packages/grafana-data/src/types/datasource.ts#L420-L432

`DataQueryResponse.data` is feed into this method and `DataQueryResponseData` can actually be `any`, but in our situations it should be `DataFrame`.

Reply:
Ah, I was confused by `DataQueryResponse.data` and `DataQueryResponseData` and thought both are the same. Disregard this!